srml/im-online/src/lib.rs

```diff
@@ -336,25 +355,56 @@ impl<T: Trait> Module<T> {
 			Some(start) => {
 				// iterate over every session
 				for index in start..curr  {
-					if <ReceivedHeartbeats<T>>::exists(&index, authority_id) {
+					let got_heartbeat = <ReceivedHeartbeats<T>>::exists(&index, authority_id);
+					let was_author = <BlockAuthors<T>>::exists(&index, authority_id);
+					if  got_heartbeat || was_author {
 						return true;
 					}
 				}
 				false
 			},
-			None => <ReceivedHeartbeats<T>>::exists(&curr, authority_id),
+			None => {
+				let got_heartbeat = <ReceivedHeartbeats<T>>::exists(&curr, authority_id);
+				let was_author = <BlockAuthors<T>>::exists(&curr, authority_id);
+				got_heartbeat || was_author
+			},
 		}
 	}
 
 	/// Returns `true` if a heartbeat has been received for `AuthorityId`
 	/// during the current session. Otherwise `false`.
 	pub fn is_online_in_current_session(authority_id: &T::AuthorityId) -> bool {
-		let current_session = <session::Module<T>>::current_index();
-		<ReceivedHeartbeats<T>>::exists(&current_session, authority_id)
+		let curr = <session::Module<T>>::current_index();
+		let got_heartbeat = <ReceivedHeartbeats<T>>::exists(&curr, authority_id);
+		let was_author = <BlockAuthors<T>>::exists(&curr, authority_id);
+		got_heartbeat || was_author
+	}
+
+	/// Returns `true` if a heartbeat has been received for `AuthorityId`
+	/// during the previous session. Otherwise `false`.
+	pub fn is_online_in_previous_session(authority_id: &T::AuthorityId) -> bool {
+		let index = <session::Module<T>>::current_index() - 1;
```

Comment:
what if `index` == 0 here? then it may wrap to u32::max_value() is that a problem?